### ISSUE TITLE
Field to constant comparison in ON clause is not working in 21.8 - No columns in one of equality side.: While processing value...

### ISSUE THREAD

**Issue by siradjev** [ClickHouse/ClickHouse#35709]
It was working in 20.8, 21.3 and is not working on 21.8. 21.9 is working again. 
It was identified in this comment as "new feature", which does not seem to be correct, since it is working since 20.8. 
https://github.com/ClickHouse/ClickHouse/issues/20193#issuecomment-949396269 #20193

There were series of fixes, but not sure which one caused the problem. 
#15162

**How to reproduce**
20.8 works
```
server-v20_8 :) select version();

SELECT version()

┌─version()─┐
│ 20.8.9.6  │
└───────────┘

1 rows in set. Elapsed: 0.044 sec.

server-v20_8 :) select t1.name, t1.value t1val, t2.value t2val
:-] from system.build_options t1 join
:-] system.build_options t2
:-] ON t1.name=t2.name AND t1.value='x86_64';

SELECT
    t1.name,
    t1.value AS t1val,
    t2.value AS t2val
FROM system.build_options AS t1
INNER JOIN system.build_options AS t2 ON (t1.name = t2.name) AND (t1.value = 'x86_64')

┌─name─────────────┬─t1val──┬─t2val──┐
│ SYSTEM_PROCESSOR │ x86_64 │ x86_64 │
└──────────────────┴────────┴────────┘

1 rows in set. Elapsed: 0.071 sec.
```
21.3 works
```
server-v21_3 :) select version();

SELECT version()

Query id: 0582c6d9-43bb-4fca-931e-84db5bac8ed3

┌─version()─┐
│ 21.3.14.1 │
└───────────┘

1 rows in set. Elapsed: 0.008 sec.

server-v21_3 :) select t1.name, t1.value t1val, t2.value t2val
:-] from system.build_options t1 join
:-] system.build_options t2
:-] ON t1.name=t2.name AND t1.value='x86_64';

SELECT
    t1.name,
    t1.value AS t1val,
    t2.value AS t2val
FROM system.build_options AS t1
INNER JOIN system.build_options AS t2 ON (t1.name = t2.name) AND (t1.value = 'x86_64')

Query id: 80635494-cc27-424b-9300-20ccbb198052

┌─name─────────────┬─t1val──┬─t2val──┐
│ SYSTEM_PROCESSOR │ x86_64 │ x86_64 │
└──────────────────┴────────┴────────┘

1 rows in set. Elapsed: 0.076 sec.

```

21.8 not working
```
server-v21_8 :) select version();

SELECT version()

Query id: 16da7163-1957-43d1-8311-ad3bdb2ac5bd

┌─version()─┐
│ 21.8.14.5 │
└───────────┘


server-v21_8 :) select t1.name, t1.value t1val, t2.value t2val
:-] from system.build_options t1 join
:-] system.build_options t2
:-] ON t1.name=t2.name AND t1.value='x86_64';

SELECT
    t1.name,
    t1.value AS t1val,
    t2.value AS t2val
FROM system.build_options AS t1
INNER JOIN system.build_options AS t2 ON (t1.name = t2.name) AND (t1.value = 'x86_64')

Query id: 28b0a18e-d792-4d6b-a3fd-e0cdfedfbd94

0 rows in set. Elapsed: 0.004 sec.

Received exception from server (version 21.8.14):
Code: 403. DB::Exception: Received from localhost:9000. DB::Exception: Not equi-join ON expression: value = 'x86_64'. No columns in one of equality side.: While processing value = 'x86_64'.
```

**Additional context**
Such query expressions are quite frequent, so 21.8.14 upgrade seem to affect a lot of deployments.

**Comment by vdimir** [comment]
So, feature don't work on 21.8, but works in next versions (21.9+), is it? 
In such cases we strongly recommend to update. Support of such queries in old versions had some issues, so, it was reverted https://github.com/ClickHouse/ClickHouse/pull/22753 and implemented in 21.9 using different approach.

**Comment by siradjev** [comment]
> So, feature don't work on 21.8, but works in next versions (21.9+), is it? In such cases we strongly recommend to update. Support of such queries in old versions had some issues, so, it was reverted #22753 and implemented in 21.9 using different approach.

Thank you for the response. 21.9 unfortunately is not an LTS version, and it will be hard for on-premise deployments to make related upgrades. 22.3 on the other hand is a bit "raw" at this moment. 

Problem is that feature was working in 20.8, and 21.3, and it's quite common usage in queries.

If it is not planned to be fixed in 21.8, it may be worth to include it in changelog with "Backward Incompatible Changes" to point the versions that have problem with such usage.

**Comment by alexey-milovidov** [comment]
Closing because the issue is fixed in the subsequent versions.
